<a href="https://colab.research.google.com/github/IfrazQazi/Book-Recommendation-System/blob/main/Book_Recommendation_system.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**BOOK RECOMMENDATION SYSTEM**
##**Problem Statement**
## During the last few decades, with the rise of Youtube, Amazon, Netflix, and many other such web services, recommender systems have taken more and more place in our lives. From e-commerce (suggest to buyers articles that could interest them) to online advertisement (suggest to users the right contents, matching their preferences), recommender systems are today unavoidable in our daily online journeys.
##In a very general way, recommender systems are algorithms aimed at suggesting relevant items to users (items being movies to watch, text to read, products to buy, or anything else depending on industries).
##Recommender systems are really critical in some industries as they can generate a huge amount of income when they are efficient or also be a way to stand out significantly from competitors. The main objective is to create a book recommendation system for users.

## Content
## The Book-Crossing dataset comprises 3 files.
##● Users
##Contains the users. Note that user IDs (User-ID) have been anonymized and map to integers. Demographic data is provided (Location, Age) if available. Otherwise, these fields contain NULL values.
## ● Books
## Books are identified by their respective ISBN. Invalid ISBNs have already been removed from the dataset. Moreover, some content-based information is given (Book-Title, Book-Author, Year-Of-Publication, Publisher), obtained from Amazon Web Services. Note that in the case of several authors, only the first is provided. URLs linking to cover images are also given, appearing in three different flavors (Image-URL-S, Image-URL-M, Image-URL-L), i.e., small, medium, large. These URLs point to the Amazon website.
##● Ratings
##Contains the book rating information. Ratings (Book-Rating) are either explicit, expressed on a scale from 1-10 (higher values denoting higher appreciation), or implicit, expressed by 0.

In [1]:
#Importing libraries
import pandas as pd
import sys
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import random
# This is to supress the warning messages (if any) generated in our code
import warnings
warnings.filterwarnings('ignore')

In [3]:
# Mounting drive with google colab notebook
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Loading data

In [6]:
## loading book data set
books = pd.read_csv('/content/drive/MyDrive/AlmaBetter/Book Recommendation system/Books.csv')

In [7]:
## loading users data set
user = pd.read_csv('/content/drive/MyDrive/AlmaBetter/Book Recommendation system/Users.csv')

In [8]:
## loading Ratings data set
rating = pd.read_csv('/content/drive/MyDrive/AlmaBetter/Book Recommendation system/Ratings.csv')

## Lets check Book data set

In [9]:
books.shape

(271360, 8)

## books data set is having shape of 271360 observation 8 columns.

In [10]:
print("Columns: ", list(books.columns))
books.head()

Columns:  ['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher', 'Image-URL-S', 'Image-URL-M', 'Image-URL-L']


ISBN  ...                                        Image-URL-L
0  0195153448  ...  http://images.amazon.com/images/P/0195153448.0...
1  0002005018  ...  http://images.amazon.com/images/P/0002005018.0...
2  0060973129  ...  http://images.amazon.com/images/P/0060973129.0...
3  0374157065  ...  http://images.amazon.com/images/P/0374157065.0...
4  0393045218  ...  http://images.amazon.com/images/P/0393045218.0...

[5 rows x 8 columns]

In [11]:
books.describe()

ISBN  ...                                        Image-URL-L
count       271360  ...                                             271357
unique      271360  ...                                             271041
top     0195153448  ...  http://images.amazon.com/images/P/225307649X.0...
freq             1  ...                                                  2

[4 rows x 8 columns]

## **Observation**
* we can see that ISBN is unique book number
* 242135 unique book Title 
* 102023 Book author 
* Year of publication  is having 202 unique years
* 16807 Publisher
* we can see that 3 columns belongs to image URL 
* Image URL S and Image URL M is having same number of unique URL but Image URL L is having less number of URL as compare to Image URL S and M

In [29]:
#dropping last three columns containing image URLs which will not be required for analysis
books.drop(['Image-URL-S', 'Image-URL-M', 'Image-URL-L'],axis=1,inplace=True)

In [30]:
## Checking for null values
books.isnull().sum() 

ISBN                   0
Book-Title             0
Book-Author            1
Year-Of-Publication    0
Publisher              2
dtype: int64

In [31]:
books.loc[books['Book-Author'].isnull(),:]

ISBN  ...                       Publisher
187689  9627982032  ...  Edinburgh Financial Publishing

[1 rows x 5 columns]

In [32]:
books.loc[books['Book-Title']==books.iloc[187689,1]]

ISBN  ...                       Publisher
187689  9627982032  ...  Edinburgh Financial Publishing

[1 rows x 5 columns]

In [33]:
books.loc[books['Publisher'].isnull(),:]

ISBN       Book-Title  ... Year-Of-Publication Publisher
128890  193169656X      Tyrant Moon  ...                2002       NaN
129037  1931696993  Finders Keepers  ...                2001       NaN

[2 rows x 5 columns]

In [36]:
books.loc[books['Book-Title']==books.iloc[128890,1]]

ISBN   Book-Title      Book-Author Year-Of-Publication Publisher
128890  193169656X  Tyrant Moon  Elaine Corvidae                2002       NaN

In [35]:
books.loc[books['Book-Title']==books.iloc[129037,1]]

ISBN  ...                          Publisher
10799   082177364X  ...                        Zebra Books
42019   0070465037  ...              McGraw-Hill Companies
58264   0688118461  ...       Harpercollins Juvenile Books
66678   1575663236  ...  Kensington Publishing Corporation
129037  1931696993  ...                                NaN
134309  0156309505  ...                      Voyager Books
173473  0973146907  ...             Red Tower Publications
195885  0061083909  ...                        HarperTorch
211874  0373261160  ...                  Worldwide Library

[9 rows x 5 columns]

In [37]:
#Filling Nan of Book-Author with others
books['Book-Author'].fillna('other',inplace=True)

In [38]:
#Filling Nan of Publisher with others
books.Publisher.fillna('other',inplace=True)

In [39]:
books.isna().sum()

ISBN                   0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
dtype: int64

In [49]:
## Checking for column Year-of-publication
books['Year-Of-Publication'].unique()

array([2002, 2001, 1991, 1999, 2000, 1993, 1996, 1988, 2004, 1998, 1994,
       2003, 1997, 1983, 1979, 1995, 1982, 1985, 1992, 1986, 1978, 1980,
       1952, 1987, 1990, 1981, 1989, 1984, 0, 1968, 1961, 1958, 1974,
       1976, 1971, 1977, 1975, 1965, 1941, 1970, 1962, 1973, 1972, 1960,
       1966, 1920, 1956, 1959, 1953, 1951, 1942, 1963, 1964, 1969, 1954,
       1950, 1967, 2005, 1957, 1940, 1937, 1955, 1946, 1936, 1930, 2011,
       1925, 1948, 1943, 1947, 1945, 1923, 2020, 1939, 1926, 1938, 2030,
       1911, 1904, 1949, 1932, 1928, 1929, 1927, 1931, 1914, 2050, 1934,
       1910, 1933, 1902, 1924, 1921, 1900, 2038, 2026, 1944, 1917, 1901,
       2010, 1908, 1906, 1935, 1806, 2021, '2000', '1995', '1999', '2004',
       '2003', '1990', '1994', '1986', '1989', '2002', '1981', '1993',
       '1983', '1982', '1976', '1991', '1977', '1998', '1992', '1996',
       '0', '1997', '2001', '1974', '1968', '1987', '1984', '1988',
       '1963', '1956', '1970', '1985', '1978', '1973', '1980'

## We can see in the Year-Of-Publication column there is some string values in it. 
##So lets check it 

In [50]:
books.loc[books['Year-Of-Publication'] == 'DK Publishing Inc',:]

ISBN  ...                                          Publisher
209538  078946697X  ...  http://images.amazon.com/images/P/078946697X.0...
221678  0789466953  ...  http://images.amazon.com/images/P/0789466953.0...

[2 rows x 5 columns]

##As it can be seen from above that there are some incorrect entries in Year-Of-Publication field. It looks like Publisher names 'DK Publishing Inc' and 'Gallimard' have been incorrectly loaded as Year-Of-Publication in dataset due to some errors in csv file